### PR TITLE
#Sql tasks. Gaziz_Tuleukulov

### PR DESCRIPTION
**Pull request by Gaziz666**
https://travis-ci.org/github/Gaziz666/db-assignments/jobs/713951958

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -105,7 +154,22 @@ async function task_1_6(db) {
  *
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT  
+            e.EmployeeID as EmployeeId, 
+            concat(e.FirstName, " ", e.LastName)  as FullName,
+            CASE
```
Comment (AlexMokin) [inline task/sql-tasks.js]: check ifnull and coalese functions

File: task/sql-tasks.js [hunk 2]
```diff
@@ -105,7 +154,22 @@ async function task_1_6(db) {
...
+        SELECT  
+            e.EmployeeID as EmployeeId, 
+            concat(e.FirstName, " ", e.LastName)  as FullName,
+            CASE
+                WHEN e.ReportsTo is null
+                THEN "-"
+                ELSE concat(r.FirstName, " ", r.LastName)
+                END
+            as ReportsTo     
+        FROM northwind.Employees as e
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not include schema name in the queries, it should be part of connection settings, Fix in all queries

File: task/sql-tasks.js [hunk 3]
```diff
@@ -117,7 +181,17 @@ async function task_1_7(db) {
  *
  */
 async function task_1_8(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            c.CategoryName, count(p.ProductName) as TotalNumberOfProducts
+        FROM 
+            northwind.Categories as c
+        RIGHT JOIN northwind.Products as p
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use right join if not necessary

File: task/sql-tasks.js [hunk 4]
```diff
@@ -177,7 +280,12 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            COUNT(ProductID) as TotalOfCurrentProducts, sum(Discontinued) as TotalOfDiscontinuedProducts
```
Comment (AlexMokin) [inline task/sql-tasks.js]: **sum(Discontinued)** - use count when you need to count records, it better shows intent

 it's better to write this query like this
notice that from-clause is optional.



```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 5]
```diff
@@ -199,7 +313,24 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            sum(month(OrderDate) = 1) as January,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it better shows intent

File: task/sql-tasks.js [hunk 6]
```diff
@@ -223,7 +360,16 @@ async function task_1_16(db) {
  *
  */
 async function task_1_17(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            c.CategoryName, avg(p.UnitPrice) as AvgPrice
+        FROM northwind.Categories as c
+        right JOIN northwind.Products as p
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use right join if not necessary. Prefer left join if possible

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 6 above)
Comment (Gaziz666) [reply task/sql-tasks.js]: fixed bugs

File: task/sql-tasks.js [hunk 7]
```diff
@@ -177,7 +275,17 @@ async function task_1_12(db) {
...
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT 
+        (SELECT COUNT(ProductID))as TotalOfCurrentProducts, 
+        (
+            SELECT COUNT(Discontinued)
+            FROM Products
+            WHERE Discontinued > 0
+        ) as TotalOfDiscontinuedProducts
+    FROM Products
```
Comment (AlexMokin) [inline task/sql-tasks.js]: from-clause is optional in this case it's better to just use 2 subqueries

File: task/sql-tasks.js [hunk 8]
```diff
@@ -199,7 +313,73 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js:319]: do not use subqueries here

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 7 above)
Comment (Gaziz666) [reply task/sql-tasks.js]: sorry i didn't understand, i have to use 2 subqueries. maybe it's wrong code formatting.

Comment (AlexMokin) [reply task/sql-tasks.js]:
```
select
   subquery1 as 'TTT'
   subquery2  as 'TTTTT'
<!--- the end -->

```

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by Gaziz666** [comment]
ok I understand, and fix mistake

**Comment by AlexMokin** [comment]
accepted

**Pull request closed without merging**